src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Blobs/BlobEncoders.cs

```diff
@@ -0,0 +1,1125 @@
+        {
+            Builder = builder;
+        }
+
+        public CustomModifiersEncoder<SignatureTypeEncoder<BlobEncoder>> FieldSignature()
+        {
+            Builder.WriteByte((byte)SignatureKind.Field);
+            return new CustomModifiersEncoder<SignatureTypeEncoder<BlobEncoder>>(
+                new SignatureTypeEncoder<BlobEncoder>(this));
+        }
+
+        public GenericTypeArgumentsEncoder<BlobEncoder> MethodSpecificationSignature(int genericArgumentCount)
+        {
+            // TODO: arg validation
+
+            Builder.WriteByte((byte)SignatureKind.MethodSpecification);
+            Builder.WriteCompressedInteger((uint)genericArgumentCount);
+
+            return new GenericTypeArgumentsEncoder<BlobEncoder>(this, genericArgumentCount);
+        }
+
+        public MethodSignatureEncoder<BlobEncoder> MethodSignature(SignatureCallingConvention convention, int genericParameterCount, bool isInstanceMethod)
+        {
+            // TODO: arg validation
+
+            var attributes = 
+                (genericParameterCount != 0 ? SignatureAttributes.Generic : 0) | 
+                (isInstanceMethod ? SignatureAttributes.Instance : 0);
+
+            Builder.WriteByte(SignatureHeader(SignatureKind.Method, convention, attributes).RawValue);
+
+            if (genericParameterCount != 0)
+            {
+                Builder.WriteCompressedInteger((uint)genericParameterCount);
+            }
+
+            return new MethodSignatureEncoder<BlobEncoder>(this, isVarArg: convention == SignatureCallingConvention.VarArgs);
+        }
+
+        public MethodSignatureEncoder<BlobEncoder> PropertySignature(bool isInstanceProperty)
+        {
+            Builder.WriteByte(SignatureHeader(SignatureKind.Property, SignatureCallingConvention.Default, (isInstanceProperty ? SignatureAttributes.Instance : 0)).RawValue);
+            return new MethodSignatureEncoder<BlobEncoder>(this, isVarArg: false);
+        }
+
+        public FixedArgumentsEncoder<NamedArgumentsEncoder<BlobEncoder>> CustomAttributeSignature(int namedArgumentCount)
+        {
+            if (unchecked((ushort)namedArgumentCount) > ushort.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namedArgumentCount));
+            }
+
+            Builder.WriteUInt16(0x0001);
+
+            return new FixedArgumentsEncoder<NamedArgumentsEncoder<BlobEncoder>>(
+                new NamedArgumentsEncoder<BlobEncoder>(this, (ushort)namedArgumentCount, writeCount: true));
+        }
+
+        public LocalVariablesEncoder<BlobEncoder> LocalVariableSignature(int count)
+        {
+            Builder.WriteByte((byte)SignatureKind.LocalVariables);
+            Builder.WriteCompressedInteger((uint)count);
+            return new LocalVariablesEncoder<BlobEncoder>(this, count);
+        }
+
+        // TODO: TypeSpec is limited to structured types (doesn't have primitive types, TypeDefRefSpec)
```

Comment:
I don't think we should treat TypeSpec as limited to the letter of the spec. modopt(anything but-type-def-or-ref) creates a primitive type-spec and roslyn already has code to explicitly allow that: http://source.roslyn.io/#Microsoft.CodeAnalysis/MetadataReader/MetadataDecoder.cs,705 

Spec claims modifier must be def-or-ref, but we'd regress that deliberate code in Roslyn and PEVerify and CLR work fine with it. 

There is another case of TypeSpec beyond what is listed in the specification: constrained prefix taking type parameter: http://stackoverflow.com/questions/7203358/rogue-entries-in-typespec-metadata-table

Pragmatically, it seems letting TypeSpec just be a way to store a signature type is best. This is how signature decoder works.